### PR TITLE
[WIP, src] Supporting BatchRenorm Under Kaldi

### PR DESCRIPTION
**Pull request by GaofengCheng**
Changes are under nnet3/nnet-normalize-component.*

### CODE REVIEW

File: src/nnet3/nnet-normalize-component.h [hunk 1]
```diff
@@ -269,6 +272,28 @@ class BatchNormComponent: public Component {
   // (smaller -> slower).  The same config exists in NormalizeComponent.
   BaseFloat target_rms_;
 
+  // the flag whether use the batch renorm described in paper:
+  // https://arxiv.org/abs/1702.03275
+  // This adds extra variables during training. The inference is the same 
+  // for either value of this parameter.
+  bool batch_renorm_;
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.h:279]: I would prefer you to do this as a separate component, it will keep the code cleaner.
Of course, you'd have to modify code about setting batchnorm test-mode, to do this type of component too.

File: src/nnet3/nnet-normalize-component.h [hunk 2]
```diff
@@ -269,6 +272,28 @@ class BatchNormComponent: public Component {
...
+  // The correction `(r, d)` is used as `corrected_value = normalized_value * r + d`,
+  // with `r` clipped to [1 / rmax, rmax], and `d` to [-dmax, dmax]. 
+  BaseFloat r_max_;
+  BaseFloat d_max_;
+
+  // Momentum used to update the moving means and standard
+  // deviations with renorm. 
+  BaseFloat renorm_momentum_;
+
+  CuVector<double> moving_stddv_;
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.h:293]: misspelled: stddev.

File: src/nnet3/nnet-normalize-component.h (hunk 1 above)
Comment (GaofengCheng) [reply src/nnet3/nnet-normalize-component.h:279]: After confirming no bug in the implementation. I can make it a separate component.

File: src/nnet3/nnet-normalize-component.cc [hunk 3]
```diff
@@ -396,6 +432,22 @@ void BatchNormComponent::InitFromConfig(ConfigLine *cfl) {
...
+  This is all dimension-by-dimension, so we just imagine the inputs
+  are scalars x(i), for i=0 .. n-1.
+
+  FORWARD PASS:
+  z(i) = batch-norm(x(i))
+  z_renorm(i) = z(i) r + d
+  where r and d are constant correction vector, for which the gradient is stopped
+  
+  BACKWARD PASS:
+  x_{renorm}'(i) = r x'(i)
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.cc:449]: This line is not correct, there should be no 'r' factor here.

File: src/nnet3/nnet-normalize-component.cc [hunk 4]
```diff
@@ -425,29 +477,90 @@ void* BatchNormComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
...
+        stddv_tmpt.ApplyPow(-1.0);
+        clipped_r.CopyFromVec(stddv_tmpt);
+        clipped_r.MulElements(moving_stddv);
+
+        clipped_r.ApplyCeiling(r_max_);
+        clipped_r.ApplyFloor(1.0 / r_max_);
+        clipped_d.ApplyCeiling(d_max_);
+        clipped_d.ApplyFloor(- d_max_);
+        
+        out->MulColsVec(clipped_r);
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.cc:557]: This is very inefficient: you are doing two multiplications on 'out' instead of one.

Comment (danpovey) [reply src/nnet3/nnet-normalize-component.cc:557]: It would be very hard for me to verify that this code is correct, because the structure is quite confusing.  If you had a separate component it would be easier to look at.

Comment (GaofengCheng) [reply src/nnet3/nnet-normalize-component.cc:557]: OK, now I'm restructuring it into a separate component

File: src/nnet3/nnet-normalize-component.cc [hunk 5]
```diff
@@ -579,11 +699,42 @@ void BatchNormComponent::StoreStats(
...
+    if (training_begining_) {
+      KALDI_WARN << "Training Begining : " << training_begining_;
+      training_begining_ = false;
+      moving_mean_.CopyFromVec(mean);
+      CuSubVector<BaseFloat> scale(memo->mean_uvar_scale, 2);
+      scale.ApplyPow(-1.0);
+      moving_stddv_.CopyFromVec(scale);
+      scale.ApplyPow(-1.0);
+    } else {
+      CuVector<BaseFloat> mv_mean(moving_mean_), mv_stddv(moving_stddv_);
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.cc:721]: these variable names are not good; e.g. moving_mean_copy_ might be better.

File: src/nnet3/nnet-normalize-component.cc [hunk 6]
```diff
@@ -579,11 +699,42 @@ void BatchNormComponent::StoreStats(
...
+
+      moving_mean_.Scale(renorm_momentum_2);
+      moving_mean_.AddVec(renorm_momentum_, mean);
+
+      CuSubVector<BaseFloat> scale(memo->mean_uvar_scale, 2);
+      scale.ApplyPow(-1.0);
+
+      moving_stddv_.Scale(renorm_momentum_2);
+      moving_stddv_.AddVec(renorm_momentum_, scale);
+      mv_mean.CopyFromVec(moving_mean_);
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.cc:733]: I think this is backwards.  mv_mean is the temporary, why are you copying from the class member?  Or maybe this is old code that's not needed now.

File: src/nnet3/nnet-normalize-component.cc [hunk 7]
```diff
@@ -606,6 +757,22 @@ void BatchNormComponent::Read(std::istream &is, bool binary) {
   stats_sumsq_.AddVecVec(1.0, stats_sum_, stats_sum_, 1.0);
   stats_sum_.Scale(count_);
   stats_sumsq_.Scale(count_);
+  ExpectToken(is, binary, "<BatchRenorm>");
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.cc:760]: this would have broken any existing BatchNormComponents on disk.

Comment (GaofengCheng) [reply src/nnet3/nnet-normalize-component.cc:760]: yes, back-level compatibility is needed here.

File: src/nnet3/nnet-normalize-component.cc [hunk 8]
```diff
@@ -221,28 +225,38 @@ void BatchNormComponent::ComputeDerived() {
     stats_sum_.SetRandn();
     stats_sumsq_.SetRandn();
     stats_sumsq_.AddVecVec(1.0, stats_sum_, stats_sum_, 1.0);
+    count_eq_zero = true;
+    // because the moving_mean_ and moving_stddv_ will not be set to zero
```
Comment (danpovey) [inline src/nnet3/nnet-normalize-component.cc:229]: Right now you are not doing anything in the Add()/Scale() functions w.r.t. moving_mean_ and moving_stddev_, so it wouldn't correctly take the average when averaging models, it would just take one of them.
Normally it's better to represent and average these things as stats, not things like mean and stddev, but it's not how the guys wrote the paper.  (That would be easier w.r.t. the Scale() function).

Comment (GaofengCheng) [reply src/nnet3/nnet-normalize-component.cc:229]: Yes, at this stage, I just pick one of the moving_mean/var of the parallel models (normally it should be the nnet[0] ). 
It's hard to do the sum and scale which we used for batchnorm for batch-renorm.
I think one way to improve is average the moving_mean/var of the parallel jobs.

Comment (GaofengCheng) [reply src/nnet3/nnet-normalize-component.cc:229]: But my experiments showed that, this way (pick moving_mean/var of one of the parallel jobs but did nothing like averaging or so) worked well. I guess the averaging is minor effect for batch-renorm.

File: src/nnet3/nnet-normalize-component.cc (hunk 3 above)
Comment (GaofengCheng) [reply src/nnet3/nnet-normalize-component.cc:449]: I checked that paper, it seems the scaler 'r' is necessary for backward pass if we scale the forward pass. Maybe I'm wrong at this point.

**Comment by danpovey** [comment]
Oh wait sorry, that factor should be there.

**Comment by GaofengCheng** [comment]
en. At this point, I'm restructuring the code to make it a separate component. Then I have to add the scripts level support for r/d schedule....

**Comment by danpovey** [comment]
BTW, I am not confident at all that your backprop code is correct.  Inside
the components the backprop is pretty much done all manually, and unless
you are comfortable with deriving (and optimizing) backprop equations, it's
going to be hard for you to figure out what the equations are.  I'm
certainly not seeing the type of comments working out the math of the
backprop, that I would have expected to see.
I'm wondering whether I should actually do it myself, and just have you
test it.  But in the meantime there is no harm in you working on it.

**Comment by GaofengCheng** [comment]
I just view the batch-norm as a black-box... just scale its forward and backward...

**Comment by GaofengCheng** [comment]
this will be replaced by a separate-component version.

**Comment by danpovey** [comment]
My concern is that the equations I used for the backprop (particularly: for
the indirect part of the derivative) may have relied on identities that are
no longer true in the case of batch renorm.  Unless you understand what I
did, you wouldn't be able to know.

**Comment by GaofengCheng** [comment]
I may have to try to understand your back-prop equations... it's different from the ones in paper. I will have a try today. If I cannot handle it, I may seek you for help.
